include/envoy/upstream/cluster_manager.h

```diff
@@ -339,6 +393,21 @@ class CdsApi {
 
 using CdsApiPtr = std::unique_ptr<CdsApi>;
 
+/**
+ * Abstract interface for a On-Demand CDS API provider.
+ */
+class OdCdsApi {
+public:
+  virtual ~OdCdsApi() = default;
+
+  /**
+   * File an on-demand request for a cluster.
+   */
+  virtual void updateOnDemand(const std::string& cluster_name) PURE;
+};
+
+using OdCdsApiPtr = std::unique_ptr<OdCdsApi>;
```

Comment:
Somewhat related to @htuch comments, but why is this public interface needed? Conceptually, something asks the cluster manager to fetch a cluster, and gets a callback when complete or failed. I don't think any of this should be exposed "externally" and should be an internal implementation detail.

Reply:
Since the config source for the on-demand discovery is provided in the per-route config, which makes it a separate object than CM has from `cds_config`, I need to pass a "CDS" object to CM. It needs to be a public interface, because it's being used in `ClusterManager` interface. I had two options here: either extend the `CdsApi` interface with the `updateOnDemand` function and pass that to CM or create a new interface (`OdCdsApi`). I went with the latter, because I found the former to be awkward. It's because `CdsApi` provides `initialize`, `setInitializedCb` and `versionInfo` and for the on-demand discovery I would use none of it.